garage/tf/algos/npo.py

```diff
@@ -153,3 +137,274 @@ def get_itr_snapshot(self, itr, samples_data):
+                )
+                pol_mean_kl = tf.reduce_mean(kl)
+
+            # Calculate surrogate loss
+            with tf.name_scope("surr_loss"):
+                lr = pol_dist.likelihood_ratio_sym(
+                    i.valid.action_var,
+                    i.valid.policy_old_dist_info_vars,
+                    policy_dist_info_valid,
+                    name="lr")
+
+                surr_vanilla = lr * adv_valid
+
+                if self._pg_loss == PGLoss.VANILLA:
+                    # VPG, TRPO use the standard surrogate objective
+                    surr_obj = tf.identity(surr_vanilla, name="surr_obj")
+                elif self._pg_loss == PGLoss.CLIP:
+                    lr_clip = tf.clip_by_value(
+                        lr,
+                        1 - self.clip_range,
+                        1 + self.clip_range,
+                        name="lr_clip")
+                    surr_clip = lr_clip * adv_valid
+                    surr_obj = tf.minimum(
+                        surr_vanilla, surr_clip, name="surr_obj")
+                else:
+                    raise NotImplementedError("Unknown PGLoss")
+
+                # Maximize E[surrogate objective] by minimizing
+                # -E_t[surrogate objective]
+                surr_loss = -tf.reduce_mean(surr_obj)
+
+            # Diagnostic functions
+            self.f_policy_kl = tensor_utils.compile_function(
+                flatten_inputs(self._policy_opt_inputs),
+                pol_mean_kl,
+                log_name="f_policy_kl")
+
+            self.f_rewards = tensor_utils.compile_function(
+                flatten_inputs(self._policy_opt_inputs),
+                rewards,
+                log_name="f_rewards")
+
+            returns = self._build_returns(rewards)
+            self.f_returns = tensor_utils.compile_function(
+                flatten_inputs(self._policy_opt_inputs),
+                returns,
+                log_name="f_returns")
+
+            return surr_loss, pol_mean_kl
+
+    def _build_entropy_terms(self, i):
+        with tf.name_scope("policy_entropy"):
+            policy_dist_info_flat = self.policy.dist_info_sym(
+                i.flat.obs_var,
+                i.flat.policy_state_info_vars,
+                name="policy_dist_info_flat")
+
+            policy_entropy_flat = self.policy._dist.entropy_sym(
+                policy_dist_info_flat)
+            policy_entropy = tf.reshape(policy_entropy_flat,
+                                        [-1, self.max_path_length])
+            policy_entropy = tf.reduce_mean(policy_entropy * i.valid_var)
+
+        self.f_policy_entropy = tensor_utils.compile_function(
+            flatten_inputs(self._policy_opt_inputs),
+            tf.reduce_mean(policy_entropy * i.valid_var),
+            log_name="f_policy_entropy")
+
+        return policy_entropy
+
+    def _build_returns(self, rewards):
+        with tf.name_scope("returns"):
+            gamma = tf.constant(
```

Comment:
this may be the same subroutine as the advantage calculation, but i recall it being slightly different.

returns are a backward-looking cumulative sum of rewards
r_i := reward at time i
R_i := return at time i
R_i = r_0 + r_1 + ... + r_i

advantages are a forward-looking sum of deltas from the baseline
A_i := advantage at time i
V_i := value function at time i
d_i := delta at time i
d_i := r_i - V_i
T := total length of trajectory

A_i = d_i + d_i+1 + ... + d_T-1 + d_T

Reply:
Yes. I have a question, why don't use `special.discount_cumsum()` here to calculate advantages and returns?

Reply:
special.discount_cumsum() is a numpy operation, so it's non-differentiable. this means we can differentiate through the advantage calculation to (for instance) an entropy bonus term.

this version does everything in tensorflow, so the entire loss function is differentiable. this means that the user can freely augment each time step of each reward with other differentiable terms, and tensorflow will automatically differentiate through them.